File: 0202-happy-number/0202-happy-number.py
class Solution:
    def genNext(self, n: int) -> int:
        total = 0

        while n > 0:
            digit = n % 10
            total += digit * digit
            n //= 10

        return total

    def isHappy(self, n: int) -> bool:
        fast = n
        slow = n

        while True:
            slow = self.genNext(slow)
            fast = self.genNext(self.genNext(fast))

            if slow == 1:
                return True

            if slow == fast:
                return False

File: 0202-happy-number/0202-happy-number.cpp
class Solution {
    int genNext(int n ){
        int sum =0;
        while(n){
            int digit = n % 10;
            sum += digit * digit;
            n = n/10;
        }
        return sum;
    }
public:
    bool isHappy(int n) {
        int fast = n;
        int slow = n;

        while(1){
            slow = genNext(slow);
            fast = genNext(genNext(fast));

            if(slow == fast){
                break;
            }
            if(slow == 1){
                break;
            }

        }

        return slow ==1;

    }
};